Changed wiki tutorial to showcase passwrd hashing with bcrypt.

mfrlin opened the pull request:
Relates to #2204

Since this tutorial does not store user data in a database I have just showcased how hashing works and explained more strongly that in real applications you are probably gong to store users in a database.

Would appreciate if someone would have read through and suggest corrections if something needs more explaining or stronger warnings about clear text dangers or something.

Disregard branch name.

mmerickel commented on docs/tutorials/wiki/authorization.rst:
```diff
@@ -61,7 +81,20 @@ request)`` returns ``None``.  We will use ``groupfinder()`` as an
 :term:`principal` or principals for a user.
 
-In a production system, user and group data will most often come from a
+There are two helper methods that will help us later when loging-in users.
```
`s/loging-in/authenticating` perhaps? I think loging-in is quite difficult to parse if it is even a real word.

mmerickel commented on docs/tutorials/wiki/src/authorization/tutorial/security.py:
```diff
@@ -1,5 +1,16 @@
+
+
+def hash_password(pw):
+    return bcrypt.hashpw(pw.encode('utf-8'), bcrypt.gensalt())
```
Just to clarify, we are aware we are returning bytes here, and happy with that? This is slightly different from the wiki2 [1] tutorial which keeps things as unicode strings at all times.

[1] https://github.com/Pylons/pyramid/blob/7ab666e3e2539662ff3324a4bdedbca3c8f8040a/docs/tutorials/wiki2/src/authentication/tutorial/models/user.py#L20-L28

stevepiercy replied:
I would like wiki and wiki2 to have consistent implementations, whichever is preferred.

mmerickel replied:
I'm not strictly against it since zodb can just handle the bytes, and the data is ascii, but I think unicode is slightly more accurate.

mmerickel commented on docs/tutorials/wiki/src/authorization/tutorial/security.py:
```diff
@@ -1,5 +1,16 @@
+    return bcrypt.hashpw(pw.encode('utf-8'), bcrypt.gensalt())
+
+def check_password(expected_hash, pw):
+    if expected_hash is not None:
```
Similarly the `expected_hash` here is expected to always be bytes, right?

stevepiercy commented on docs/tutorials/wiki/authorization.rst:
```diff
@@ -61,7 +81,20 @@ request)`` returns ``None``.  We will use ``groupfinder()`` as an
+record. If the two hashed values match, then the submitted
+password is valid, and we can authenticate the user.
+
+We hash passwords so that it is impossible to decrypt them and use them to
```
Drop extra "them" after decrypt:
We hash passwords so that it is impossible to decrypt and use them to

stevepiercy replied:
Drop extra "them" after decrypt:
We hash passwords so that it is impossible to decrypt and use them to

stevepiercy commented:
@mfrlin thanks for taking this on! Please let me know when it is ready for a human to run through manually, and I'll do that task to check your work, making sure docs align with code and other details.

mfrlin commented:
Ok, I've made implementation of helper functions consistent with wiki2 example. My reasoning was to make things simpler because there was no reason to convert hashes back to unicode.

I've also fixed wording with changes you suggested.

Added footnote from wiki2 example and a link to bcrypt.

Line numbers and highlighting are all good from what I can tell.

I've also had concern about bcrypt because it did not install on my system until I installed libffi-dev package. But I think installation is out of scope of this tutorial and footnote kinda covers it?

mmerickel commented:
Thank you @mfrlin!

mmerickel commented:
@mfrlin What system are you having issues installing bcrypt on? It has a manylinux1 wheel so it should work fine on almost all systems unless you are not using a recent version of pip. For example buildout or `setup.py develop`.

mfrlin commented:
@mmerickel I'm using Ubuntu 15.10, so maybe I am somewhat behind the times. Can't test on other systems currently.

ztane commented:
@mfrlin note that ubuntu 15.10 was eoled long ago already.
